jacodb-taint-configuration/src/main/kotlin/org/jacodb/taint/configuration/TaintConfigurationFeature.kt

```diff
@@ -295,31 +293,49 @@ class TaintConfigurationFeature private constructor(
                 return mkOr(types.flatMap { type -> position.map { TypeMatches(it, type) } })
             }
 
-            typeMatcher as ClassMatcher
+            val typeMatchers = (typeMatcher as ClassMatcher).extractAlternatives()
+            val unresolvedMatchers = mutableListOf<ClassMatcher>()
+            val disjuncts = mutableListOf<Condition>()
 
-            val pkgMatcher = typeMatcher.pkg
-            val clsMatcher = typeMatcher.classNameMatcher
             val cp = method.enclosingClass.classpath
 
-            if (pkgMatcher is NameExactMatcher && clsMatcher is NameExactMatcher) {
+            for (matcher in typeMatchers) {
+                val pkgMatcher = matcher.pkg
+                val clsMatcher = matcher.classNameMatcher
+
+                if (pkgMatcher !is NameExactMatcher || clsMatcher !is NameExactMatcher) {
+                    unresolvedMatchers += matcher
+                    continue
+                }
+
                 val type = cp.findTypeOrNull("${pkgMatcher.name}$DOT_DELIMITER${clsMatcher.name}")
-                    ?: return mkOr(emptyList())
-                return mkOr(position.map { TypeMatches(it, type) })
-            }
+                    ?: continue
 
-            val alternatives = typeMatcher.extractAlternatives()
-            val disjuncts = mutableListOf<Condition>()
+                position.mapTo(disjuncts) { TypeMatches(it, type) }
+            }
 
-            alternatives.forEach { classMatcher ->
-                val allClasses = runBlocking {
-                    cp.hierarchyExt().findSubClasses(cp.objectClass, allHierarchy = true, includeOwn = true)
+            if (unresolvedMatchers.isNotEmpty()) {
+                val allClassNames = cp.registeredLocations.flatMapTo(hashSetOf()) {
+                    val names = it.jcLocation?.classNames ?: return@flatMapTo emptyList()
```

Comment:
This api was not designed for often use so it's better at least to cache it somehow if it's possible